annnufan/assignment3/founder.cpp

```diff
@@ -0,0 +1,74 @@
+#include "dictionary.h"
+#include <set>
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <iostream>
+
+class founder {
+	dictionary dict;
+	std::vector<std::vector<char>> grid;
+	std::set<std::string> words_from_grid;
+	bool find_before;
+	int n, m;
+	std::vector<std::vector<int>> moves;
+
+	void go_from_cell(int i, int j, std::vector<std::vector<bool>> used, std::string word) {

```

Comment:
Here you pass `used` and `word` by value, which means that they are copied for each recursive call. It can greatly increase the complexity of the solution, especially if the grid is large.

An alternative which is more efficient is to have one copy for each of `used` and `word`, and pass it by pointer/reference to each recursive call.
If you choose to do so, then you need to remember to update them before and after each recursive call:
```C++
// At the beginning:
word.push_back(grid[i][j]);
used[i][j] = true;
......
// Before returning:
word.pop_back();
used[i][j] = false;
```

This solution is more efficient (no copies of large vectors for each function call), but can be less robust - forgetting to set `word` and `used` back can cause bugs. So be aware of both when choosing your solution.

Reply:
Yeah, I thought about both options and choose more readable option for this case withount "hard" clear and update after usage.